Jeffswenson ldr optimisation integration

jeffswenson opened the pull request.

michae2 committed:
opt: add swap flag to MutationPrivate

Inside the optimizer we can simply use a boolean flag to distinguish
between update and delete, and update swap and delete swap,
respectively. Add this flag to MutationPrivate and use it in execbuild
to construct the swap nodes.

Informs: #144503

Release note: None

opt/norm: add UseSwapMutation normalization rule

Add a new normalization rule which converts Update and Delete
expressions to UpdateSwap and DeleteSwap, respectively. Crucially, this
normalization rule also replace the initial Scan of the mutation with a
single-row Values expression, which optimistically provides the complete
old row (if it exists and matches the filters). This allows the
UpdateSwap and DeleteSwap operators to use CPut to validate the
existence of the old row in the same batch as modifying it, changing the
mutation statement from 2 RTs to 1 RT.

With this normalization rule in place, UpdateSwap and DeleteSwap should
be available for use by any qualifying UPDATE and DELETE statements.

Fixes: #144503
Fixes: #85328

Release note (performance improvement): Add new `update swap` and
`delete swap` operators which allow some UPDATE and DELETE statements to
execute in 1 round trip instead of 2 round trips. For now, these
operators can be used when:

- all columns in the primary index are constrained to a single exact value
  by the WHERE clause
- only a single row is modified
- there are no FK checks or cascades
- there are no uniqueness checks
- there are no check constraints
- there are no after triggers
- the table only uses a single column family

We may lift some of these restrictions in future releases.

logictest: add testcase for incorrect dupe key from UpdateSwap

In #145994 we discovered a problem with batch ordering that could cause
the wrong ConditionFailed error to be returned. This could cause
UpdateSwap and DeleteSwap to incorrectly return a duplicate key error
instead of simply modifying 0 rows.

This problem was fixed by #147117 and #147510. Now that we have
UpdateSwap working, add a testcase that verifies the problem is fixed
for UpdateSwap.

Informs: #144503, #145994

Release note: None

sql, opt: add use_swap_mutations session variable

Release note (sql change): Add new session variable `use_swap_mutations`
which controls whether the new update swap and delete swap operators are
enabled for use by UPDATE and DELETE statements.

jeffswenson committed:
logical: use isql.Session in the crud writer

logical: crud writer tweaks

logical: improve type handling in the crud handler

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/151353)
<!-- Reviewable:end -->

cockroach-teamcity commented:
<details><summary><strong>⚪ Sysbench</strong> [SQL, 3node, oltp_read_write]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 7.932m ±0% | 7.954m ±0% | ~ | p=0.061 n=15    |
| ⚪ **allocs/op** | 7.822k ±0% | 7.815k ±2% | ~ | p=0.846 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/15f46c9/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/15f46c99051aa3600d0dc478071fea1e5c5cc497/bin/pkg_sql_tests benchdiff/15f46c9/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/15f46c9/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/8f413be/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/8f413beddfe8ba1b7407a3b4f5f5df5b32f85402/bin/pkg_sql_tests benchdiff/8f413be/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/8f413be/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/SQL/3node/oltp_read_write$ --old=8f413be --new=15f46c9 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>🔴 Sysbench</strong> [KV, 3node, oltp_read_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| 🔴 **sec/op** | 1.067m ±0% | 1.078m ±1% | +1.09% | p=0.000 n=15    |
| ⚪ **allocs/op** | 1.868k ±0% | 1.869k ±0% | ~ | p=0.073 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/15f46c9/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/15f46c99051aa3600d0dc478071fea1e5c5cc497/bin/pkg_sql_tests benchdiff/15f46c9/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/15f46c9/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/8f413be/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/8f413beddfe8ba1b7407a3b4f5f5df5b32f85402/bin/pkg_sql_tests benchdiff/8f413be/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/8f413be/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_read_only$ --old=8f413be --new=15f46c9 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_write_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 1.909m ±0% | 1.913m ±0% | ~ | p=0.187 n=15    |
| ⚪ **allocs/op** | 3.793k ±0% | 3.796k ±0% | ~ | p=0.065 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/15f46c9/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/15f46c99051aa3600d0dc478071fea1e5c5cc497/bin/pkg_sql_tests benchdiff/15f46c9/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/15f46c9/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/8f413be/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/8f413beddfe8ba1b7407a3b4f5f5df5b32f85402/bin/pkg_sql_tests benchdiff/8f413be/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/8f413be/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_write_only$ --old=8f413be --new=15f46c9 ./pkg/sql/tests
```

</details>

</details>

<details><summary>Artifacts</summary>

**download**:
```shell
mkdir -p new
gcloud storage cp gs://cockroach-microbench-ci/artifacts/15f46c99051aa3600d0dc478071fea1e5c5cc497/18013134364-1/\* new/
mkdir -p old
gcloud storage cp gs://cockroach-microbench-ci/artifacts/8f413beddfe8ba1b7407a3b4f5f5df5b32f85402/18013134364-1/\* old/
```

</details>

_built with commit: [15f46c99051aa3600d0dc478071fea1e5c5cc497](https://github.com/cockroachdb/cockroach/commit/15f46c99051aa3600d0dc478071fea1e5c5cc497)_

<!-- microbench-ci -->

cockroach-teamcity commented:
<details><summary><strong>⚪ Sysbench</strong> [SQL, 3node, oltp_read_write]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 7.932m ±0% | 7.963m ±0% | +0.39% | p=0.000 n=15    |
| ⚪ **allocs/op** | 7.814k ±1% | 7.831k ±2% | ~ | p=0.083 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/9748025/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/9748025ca82afd6559b3656172e87ea0122a3aca/bin/pkg_sql_tests benchdiff/9748025/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/9748025/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/8f413be/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/8f413beddfe8ba1b7407a3b4f5f5df5b32f85402/bin/pkg_sql_tests benchdiff/8f413be/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/8f413be/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/SQL/3node/oltp_read_write$ --old=8f413be --new=9748025 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_read_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 1.084m ±1% | 1.091m ±1% | ~ | p=0.045 n=15    |
| ⚪ **allocs/op** | 1.868k ±0% | 1.869k ±0% | ~ | p=0.703 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/9748025/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/9748025ca82afd6559b3656172e87ea0122a3aca/bin/pkg_sql_tests benchdiff/9748025/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/9748025/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/8f413be/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/8f413beddfe8ba1b7407a3b4f5f5df5b32f85402/bin/pkg_sql_tests benchdiff/8f413be/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/8f413be/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_read_only$ --old=8f413be --new=9748025 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_write_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 1.928m ±1% | 1.925m ±0% | ~ | p=0.486 n=15    |
| ⚪ **allocs/op** | 3.794k ±0% | 3.794k ±0% | ~ | p=0.751 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/9748025/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/9748025ca82afd6559b3656172e87ea0122a3aca/bin/pkg_sql_tests benchdiff/9748025/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/9748025/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/8f413be/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/8f413beddfe8ba1b7407a3b4f5f5df5b32f85402/bin/pkg_sql_tests benchdiff/8f413be/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/8f413be/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_write_only$ --old=8f413be --new=9748025 ./pkg/sql/tests
```

</details>

</details>

<details><summary>Artifacts</summary>

**download**:
```shell
mkdir -p new
gcloud storage cp gs://cockroach-microbench-ci/artifacts/9748025ca82afd6559b3656172e87ea0122a3aca/18017046796-1/\* new/
mkdir -p old
gcloud storage cp gs://cockroach-microbench-ci/artifacts/8f413beddfe8ba1b7407a3b4f5f5df5b32f85402/18017046796-1/\* old/
```

</details>

_built with commit: [9748025ca82afd6559b3656172e87ea0122a3aca](https://github.com/cockroachdb/cockroach/commit/9748025ca82afd6559b3656172e87ea0122a3aca)_

<!-- microbench-ci -->

jeffswenson committed:
sessionmutator: extract sessionDataMutator into a package

This extracts the sessionDataMutator into its own package so that it can
be exposed on the isql.Session interface. I tried pushing it into the
pre-existing sessiondata package, but that creates a circular dependency
between the sessiondata and tree package.

Release note: none
Epic: CRDB-48647

jeffswenson committed:
isession: add a helper for mutating the session

This adds a MutateSession method that allows modifying the session
without going through the SQL layer. It is possible to modify the
session using `SET ...` statements, but LDR must mutate the session
before every statement and modifying the session consumes ~10% of all
CPU cycles in a cluster that is using the crud writer to replicate data.

An alternative to adding MutateSession is we could optimize the Set
statement. The main reasons its inefficient for LDR are:
1. It only supports strings, so we can set the hlc origin timestamp
	 without roundtripping through a string.
2. The optimizer does not support generic query plans for prepared
	 statements that include a set statement.

Release note: none
Epic:	CRDB-48647

cockroach-teamcity commented:
<details><summary><strong>⚪ Sysbench</strong> [SQL, 3node, oltp_read_write]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 12.81m ±2% | 12.96m ±3% | ~ | p=0.806 n=15    |
| ⚪ **allocs/op** | 8.309k ±0% | 8.305k ±0% | ~ | p=0.959 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/4e4c016/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/4e4c0162c0298d17ade2dcc67ebfd7884f3aad14/bin/pkg_sql_tests benchdiff/4e4c016/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/4e4c016/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/663fd06/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/663fd06df0f440fead6c34ff4e3a5173b94bfd57/bin/pkg_sql_tests benchdiff/663fd06/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/663fd06/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/SQL/3node/oltp_read_write$ --old=663fd06 --new=4e4c016 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_read_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 3.454m ±3% | 3.446m ±3% | ~ | p=0.486 n=15    |
| ⚪ **allocs/op** | 2.075k ±1% | 2.065k ±1% | ~ | p=0.323 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/4e4c016/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/4e4c0162c0298d17ade2dcc67ebfd7884f3aad14/bin/pkg_sql_tests benchdiff/4e4c016/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/4e4c016/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/663fd06/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/663fd06df0f440fead6c34ff4e3a5173b94bfd57/bin/pkg_sql_tests benchdiff/663fd06/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/663fd06/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_read_only$ --old=663fd06 --new=4e4c016 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_write_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 3.968m ±2% | 3.984m ±2% | ~ | p=0.935 n=15    |
| ⚪ **allocs/op** | 4.189k ±0% | 4.190k ±0% | ~ | p=0.878 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/4e4c016/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/4e4c0162c0298d17ade2dcc67ebfd7884f3aad14/bin/pkg_sql_tests benchdiff/4e4c016/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/4e4c016/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/663fd06/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/663fd06df0f440fead6c34ff4e3a5173b94bfd57/bin/pkg_sql_tests benchdiff/663fd06/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/663fd06/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_write_only$ --old=663fd06 --new=4e4c016 ./pkg/sql/tests
```

</details>

</details>

<details><summary>Artifacts</summary>

**download**:
```shell
mkdir -p new
gcloud storage cp gs://cockroach-microbench-ci/artifacts/4e4c0162c0298d17ade2dcc67ebfd7884f3aad14/18762696403-1/\* new/
mkdir -p old
gcloud storage cp gs://cockroach-microbench-ci/artifacts/663fd06df0f440fead6c34ff4e3a5173b94bfd57/18762696403-1/\* old/
```

</details>

_built with commit: [4e4c0162c0298d17ade2dcc67ebfd7884f3aad14](https://github.com/cockroachdb/cockroach/commit/4e4c0162c0298d17ade2dcc67ebfd7884f3aad14)_

<!-- microbench-ci -->
